File: dart/lib/src/webhook.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// Verifies a webhook payload using HMAC-SHA256.
class WebhookVerifier {
  /// Verifies the [payload] against the [signature] using the [secret].
  ///
  /// Returns the parsed JSON map on success.
  /// Throws [ArgumentError] for invalid inputs.
  /// Throws [FormatException] for signature mismatch.
  static Map<String, dynamic> verify(
    String payload,
    String signature,
    String secret,
  ) {
    if (payload.isEmpty) {
      throw ArgumentError('Empty webhook payload');
    }
    if (signature.isEmpty) {
      throw ArgumentError('Empty webhook signature');
    }
    if (secret.isEmpty) {
      throw ArgumentError('Empty webhook secret');
    }

    final key = utf8.encode(secret);
    final bytes = utf8.encode(payload);
    final hmacSha256 = Hmac(sha256, key);
    final digest = hmacSha256.convert(bytes);
    final expected = digest.toString();

    if (expected != signature) {
      throw const FormatException('Invalid webhook signature');
    }

    return jsonDecode(payload) as Map<String, dynamic>;
  }
}

File: go/webhook.go
package authon

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
)

// WebhookService handles webhook verification.
type WebhookService struct{}

// Verify validates a webhook payload against the provided HMAC-SHA256 signature.
// Returns the parsed event data on success.
func (w *WebhookService) Verify(payload []byte, signature string, secret string) (map[string]any, error) {
	if len(payload) == 0 {
		return nil, errors.New("authon: empty webhook payload")
	}
	if signature == "" {
		return nil, errors.New("authon: empty webhook signature")
	}
	if secret == "" {
		return nil, errors.New("authon: empty webhook secret")
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(payload)
	expected := hex.EncodeToString(mac.Sum(nil))

	if !hmac.Equal([]byte(expected), []byte(signature)) {
		return nil, errors.New("authon: invalid webhook signature")
	}

	var result map[string]any
	if err := json.Unmarshal(payload, &result); err != nil {
		return nil, errors.New("authon: failed to parse webhook payload: " + err.Error())
	}

	return result, nil
}
